### ISSUE TITLE
Provide a C library for qrexec

### ISSUE THREAD

**Issue by DemiMarie** [QubesOS/qubes-issues#4685]
### Qubes OS version:

`Qubes release 4.0 (R4.0)`

### Affected component(s): 

domU qrexec

---

### Steps to reproduce the behavior:

Look for a C library that implements qrexec

### Expected behavior:

Qubes has such a library, and the qrexec client tools wrap it.

### Actual behavior:

No such library.

### General notes:

This would be quite useful for those who want to call qrexec services from their applications.

For maximum utility, the library should be fully asynchronous.

---

**Comment by marmarek** [comment]
We're not planning to implement such library, but if you really like, we'd accept good quality patches for it. 
In the meantime, it's easy enough to handle qrexec calls using separate process and communicate with it using pipes. There surely is some performance impact from such approach, but there are much easier ways to optimize qrexec performance, starting with increasing [data chunk size](https://github.com/QubesOS/qubes-linux-utils/blob/master/qrexec-lib/qrexec.h#L28) and optimizing qrexec policy handling. https://github.com/QubesOS/qubes-issues/issues/4686 is also something worth working on (and in fact would be very useful for decent Admin API performance).

> For maximum utility, the library should be fully asynchronous.

I find it especially painful in C libraries. Every single one I've seen uses different API, frequently encouraging "callback hell".

**Comment by DemiMarie** [comment]
I agree, and would also add “increase the size of the pipe buffers in the kernel” to the list (it is just a single system call).

**Comment by DemiMarie** [comment]
@marmarek Is the qrexec protocol considered stable?  That is, is implementing it oneself supported?

**Comment by cfcs** [comment]
I hope so, we're building things on top of it. :)

**Comment by marmarek** [comment]
Qrexec protocol [have version negotiation as part of handshake](https://www.qubes-os.org/doc/qrexec3/#qrexec-protocol-details). Recently made it really negotiation, not only "gracefully reject mismatching version". So, yes, the protocol is stable.
